akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala

```diff
@@ -617,19 +621,45 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
 
   private def runInboundOrdinaryMessagesStream(compression: InboundCompressions): Unit = {
     val completed =
-      if (remoteSettings.TestMode) {
-        val (mgmt, c) = aeronSource(ordinaryStreamId, envelopePool)
-          .via(inboundFlow(compression))
-          .viaMat(inboundTestFlow)(Keep.right)
-          .toMat(inboundSink)(Keep.both)
-          .run()(materializer)
-        testStages.add(mgmt)
-        c
+      if (inboundLanes == 1) {
+        if (remoteSettings.TestMode) {
+          val (mgmt, c) = aeronSource(ordinaryStreamId, envelopeBufferPool)
+            .via(inboundFlow(compression))
+            .viaMat(inboundTestFlow)(Keep.right)
+            .toMat(inboundSink(envelopeBufferPool))(Keep.both)
+            .run()(materializer)
+          testStages.add(mgmt)
+          c
+        } else {
+          aeronSource(ordinaryStreamId, envelopeBufferPool)
+            .via(inboundFlow(compression))
+            .toMat(inboundSink(envelopeBufferPool))(Keep.right)
+            .run()(materializer)
+        }
+
       } else {
-        aeronSource(ordinaryStreamId, envelopePool)
+        val source = aeronSource(ordinaryStreamId, envelopeBufferPool)
           .via(inboundFlow(compression))
-          .toMat(inboundSink)(Keep.right)
-          .run()(materializer)
+          .watchTermination()(Keep.right)
+
+        val lane = inboundSink(envelopeBufferPool).async
+        val partitionFun: InboundEnvelope ⇒ Int = {
+          envelope ⇒
+            envelope.recipient match {
+              case OptionVal.Some(r) ⇒ math.abs(r.path.uid) % inboundLanes
```

Comment:
Document that we use this effectively consistent hashing trick to select the lanes here?
Easy to miss if not aware of it.